Refactored core model classes and support for collections of model in…

david-waltermire opened the pull request:
# Committer Notes
This PR does two things:
1. Adds support for defining a Metaschema module's model as a Metaschema module.
2. Defines a new constraints format that can target based on a Metapth against the Metaschema module as a Metaschema module.

This involves adding support for usnistgov/metaschema#228, which means this PR depends on usnistgov/metaschema#470, which addresses this issue.

To full support the features of usnistgov/metaschema#228, the following needs to be completed.

- [x] core: Add core XML model support for `choice-group`
- [x] core: Add Metapath support for choice groups
- [x] databind: Add code generation support for `choice-group`
- databind: Add XML serialization/deserialization support for `choice-group`
   - [x] serialization
   - [x] deserialization
- databind: Add JSON/YAML serialization/deserialization support for `choice-group`
   - [x] serialization
   - [x] deserialization
- [x] databind: Schema loader support
- [x] schema-gen: Add XML schema generation support  for `choice-group`
- [x] schema-gen: Add JSON schema generation support  for `choice-group`

As a stretch goal, it would be nice to add support for binding customization of the item type used   for `choice-group`, to allow a base class or interface to be used instead of `Object` as the item item. This base class or interface would need to be implemented by each member of the  `choice-group`.

- [ ] Add support for `choice-group` binding customization

### All Submissions:

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

Commit:
Refactored core model classes and support for collections of model instances.

Commit:
Relocated class files.

aj-stein-nist commented:
> This PR does two things:
> 
>     1. Adds support for defining a Metaschema module's model as a Metaschema module.
> 
>     2. Defines a new constraints format that can target based on a Metapth against the Metaschema module as a Metaschema module.

And so it begins! I look forward to see this PR progress and reach it's final state. 😎

Commit:
Refactored IBindingContext to move supporting interfaces to be child interfaces.
Removed the need to instances of IDataTypeHandler. Now IClassDataTypeHandler functionality is provided by IFeatureComplexItemValueHandler and JavaTypeAdapterDataTypeHandler is provided by IFeatureScalarItemValueHandler.
Refactored bound class hierarchy to be more consistent with needed methods.
- Moved Java field handling to IFeatureJavaField from IBoundNamedModelInstance
- Moved JSON key and item handling to IBoundModelInstance.
- Reworked XML and JSON reading and writing.
- Added IBoundChoiceGroupInstance interface for choice groups.

Commit:
Refactored ILet statement construction to use static factory methods.
Cleaned up IModelPropertyInfo and implementations, relocating supporting classes as child classes.
Implemented a IModelInstance item parsing as IModelPropertyInfo.IReadHandler, allowing all item parsing to be implemented in the format-specific parsing classes, i.e. MetaschemaJsonReader, MetaschemaXmlReader.

Commit:
Cleaned up IModelPropertyInfo readItems methods, which now return the parsed value. Eliminated the need for IPropertyCollector instances.

Commit:
Renamed many instance implementation classes to ensure consistent use of terminology. Prior the terms "property" and "instance" were used interchangeably, which could cause some confusion.

Commit:
Refactored XML and JSON item writing to use a shared IModelInstanceCollectionInfo.IWriteHandler implementation, allowing all item writing to be implemented in the format-specific classes, i.e. MetaschemaJsonWriter, MetaschemaXmlWriter.

Commit:
Cleaned up unneeded methods.

Commit:
Added some Javadocs.
Aligned annotations with class generation, adding support for many missing annotation entries.

Commit:
Fixed default value handling which wasn't producing empty collection for unparsed model instances.
Refactored JSON key handling to determine the JSON key on a per item basis. This is needed for ChoiceGroups.
Fixed IBindingMatcher to use the correct class type.

Commit:
Added binding support for choice groups.
Fixed some PMD errors.

Commit:
Moved JSON reading and writing code from IItemValueHandler to MetaschemaJsonReader and MetaschemaJsonWriter using dispatch methods. This allows are IO code to be implemented withing the JSON format-specific implementation.

Commit:
Moved XML reading and writing code from IItemValueHandler to MetaschemaXmlReader and MetaschemaXmlWriter using dispatch methods. This allows are IO code to be implemented withing the XML format-specific implementation.

Commit:
Fixed Javadocs with invalid parameters causing the build to break.

Commit:
Refactored instance class hierarchy to support grouped and ungrouped model instances, with minimal implementation.
Cleaned up getDefaultValue and getEffectiveDefaultValue methods.
Implemented choice group writing for XML, JSON, and YAML.
Initial commit of the Metaschema module bindings.

Commit:
Cleaned up IBoundGroupedFieldInstance implementations to use only the complex form, since a grouped item must have a bound class to determine its item type.
Cleaned up JSON value key naming code.
Adjusted the name used during code generation to use the value key name if available. This provides a more sensible name for Java property get and set methods created during code generation.
Added support for deprecated on allowed value enumerations.

david-waltermire commented:
> > This PR does two things:
> > ```
> > 1. Adds support for defining a Metaschema module's model as a Metaschema module.
> > 
> > 2. Defines a new constraints format that can target based on a Metapth against the Metaschema module as a Metaschema module.
> > ```
> 
> And so it begins! I look forward to see this PR progress and reach it's final state. 😎

Using this code I was able to produce an [example](https://gist.github.com/david-waltermire-nist/988ecfa6c8c2e75b5f3eec7e01739bfa) Metaschema module in YAML. This module describes a Metaschema module, so it is self describing!

Commit:
Performed some code cleanup and formatting.
Fixed the compile error in SimpleFieldInstance by creating the ISimpleScalarItemValueJavaField interface to combine the interfaces used and point the getValue and setValue methods to the correct implementation.

Commit:
Fixed bug causing grouping discriminators from being generated.

Commit:
Fixed ordering of generated flag fields. They will now be ordered in the order in which they are declared in the Metaschema module.
Fixed compile errors.

Commit:
Renamed model interfaces for greater consistency.
Cleaned up bound module loading in DefaultBindingContext.
Stubbed out grouped instance implementations.
Started refactor of choice group instances. Still need to work on Metapath items and other unimplemented methods.
Added testing of assemblies with various types of fields and flags.
Refactored DefaultJsonDeserializer to read a root element without the need to produce a Metapath item first.
Implemented new unit tests to test the model bindings.
Adjusted the class generation for grouped instances to allow for extended classes where definitions are used by reference.
Implemented subclass generation for choice group instances.
Updated generated Metaschema module classes.

Commit:
Flattened binding subclasses.

Commit:
Removed "binding" interfaces and classes, flattening these into the related "bound" interfaces and classes.

Commit:
Cleaned up XML parsing code. Fixed a bunch of bugs causing parsing to not work.

Commit:
Tuned up the Metaschema module bound classes.

Commit:
Fixed all identified bugs in XML parsing.
Added support for parsing XML fields as a root element.

Commit:
Completed JSON writer.

Commit:
Some progress towards XML writing. Need to implement choice groups still.
Fixed miscellaneous compile errors in other modules.

Commit:
Fixed issues with the XML reader.
Fixed compile errors for other modules to align with recent changes.

Commit:
Fixed remaining unit test failures.
Fixed some compile issues.
Fixed remaining unit tests.

Commit:
Added support for choice group JSON reading, but encountered a bug in the parser use.

Commit:
Added object mapper configuration to JSON factory.

Commit:
Fixed recursive parsing bug.

Commit:
fixed broken javadoc.

Commit:
Fixed a bunch of PMD and compile errors and warnings.

Commit:
Moved Metaschema module binding classes to a subdirectory to help with change integration.
Started work on Metaschema module wrappers for bound objects.

Commit:
Updated Metaschema module bindings.

Commit:
Added support for Metaschema module loading, minus the loader.

Commit:
Completed work on a Metaschema module loader that uses the Metaschema module binding instead of the XML parser. Metaschema modules in XML, JSON, and YAML can now be fully read!

Commit:
Added Metaschema module loader support for entity resolution. Tested with OSCAL Metaschema modules.

Commit:
Refactored container handling code to reuse the core model implementation where possible.
Refactored group-as namespace handling to allow for no namespace situations. Implemented getEffectiveGroupAsNamespace to handle the default case of using the module namespace where no group-as support is provided.
Relocated the IModulePostProcessor interface to the new IModuleLoader interface.
Refactored flag and field name handling for a cleaner implementation. This new approach uses "name" as the binding annotation value for flags, instead of "useName".
Completed binding support. All current JUnit tests run without error.
Added support for a model(node) Metapath function that gets a nodeset based on the underlying Metaschema module model for the node.

Commit:
Refactored core model interfaces to reduce inheritance conflicts.
Fixed remaining JUnit and schema generation bugs.

Commit:
Adjusted the module interfaces to promote a cleaner, non-parameterized API.

Commit:
Fixed compile, PMD and Spotbug warnings.

Commit:
Performed some overall interface cleanup.
Refactored JSON schema generator. Many improvements to the indexing approach used to determine which definitions become global definitions in the JSON schema.
Fixed missing properties in choice group items.
Improved JSON documentation to include titles and descriptions on properties.
JSON and XML schema generation for choice groups is working.

Commit:
Fixed some PMD and Javadoc errors.

Commit:
Fixed a bug causing constraint targets to not get parsed and instead end up as null.
Fixed a bug causing empty annotation documentation in XML for enumerated values with no description.

Commit:
Cleaned up some PMD and compile warnings.

david-waltermire commented:
Support for Metapath using choice groups is partially implemented, but not tested. I am going to work on this in a new PR, so this one can be merged as-is.
